x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/IndexingStateProcessor.java

```diff
@@ -98,11 +110,20 @@ private BytesReference splitAndPersist(BytesReference bytesRef, int searchFrom)
         return bytesRef.slice(splitFrom, bytesRef.length() - splitFrom);
     }
 
-    void persist(BytesReference bytes) throws IOException {
-        BulkRequest bulkRequest = new BulkRequest();
-        bulkRequest.add(bytes, AnomalyDetectorsIndex.jobStateIndexWriteAlias(), XContentType.JSON);
+    void findAppropriateIndexAndPersist(BytesReference bytes) throws IOException {
```

Comment:
Please can you add a comment here explaining why we have to do this.  Today everyone in the team understands what an "appropriate" index is, but it won't be so obvious 2 years from now.

Key points:

- Some types of state, for example data frame analytics state and categorizer state, are written multiple times with the same document ID
- Other types of state, for example anomaly detector state, are written with new document IDs each time
- The code needs to work correctly when the state index is rolling over
- The approach of having duplicate IDs in different state indices is tricky when we deliberately have no mappings on the state index, as we cannot sort and filter in a search
- The state documents are large, so having dead documents with duplicate IDs is not ideal from a disk usage perspective either

Possibly this could go in the top level class Javadoc comment and this method's comment could just say "for what is meant by appropriate see the class documentation".